### PR TITLE
Add support for AMD K8 temperature reporting. 

### LINKED ISSUE #128: Temperature reporting does not work on older AMD K8 CPUs
Temperature reporting does not work on older AMD K8 CPUs

So far I have tested the following CPUs that don't work:
 - AMD Athlon(tm) Dual Core Processor 5400B
 - AMD Athlon(tm) 64 X2 Dual Core Processor 5000+

Both of them are family: 0xf, model: 0x6b, stepping: 0x2

image

### PR DESCRIPTION
**Pull request by x86fr**
Add handle for various AMD temp sensors erratas. Closes #128

Tested on various K8/K10 platforms but requires more tests.

### CODE REVIEW

File: system/temperature.c [hunk 1]
```diff
@@ -16,18 +17,32 @@
...
+float cpu_temp_offset = 0;
+
 //------------------------------------------------------------------------------
 // Public Functions
 //------------------------------------------------------------------------------
 
-int get_cpu_temperature(void)
+void temperature_init(void)
 {
+    // Process SMP Quirks
```
Comment (x86fr) [inline system/temperature.c:38]: NTS> Correct copy/pasted description

File: system/hwquirks.c [hunk 2]
```diff
@@ -115,4 +150,53 @@ void quirks_init(void)
...
     }
+
+    //  ------------------------------------------------------
+    //  -- Early AMD K8 doesn't support temperature reading --
+    //  ------------------------------------------------------
+    // The on-die temperature diode on SH-B0/B3 stepping does not work.
+    if (cpuid_info.vendor_id.str[0] == 'A' && cpuid_info.version.family == 0xF
+        && cpuid_info.version.extendedFamily == 0 && cpuid_info.version.extendedModel == 0) {   // Early K8
+        if ((cpuid_info.version.model == 4 && cpuid_info.version.stepping == 0) ||              // SH-B0 ClawHammer (Athlon 64)
+            (cpuid_info.version.model == 5 && cpuid_info.version.stepping <= 1)) {              // SH-B0/B3 SledgeHmmer (Opteron)
```
Comment (debrouxl) [inline system/hwquirks.c:161]: Typo "Sledgehmmer"

File: system/temperature.c [hunk 3]
```diff
@@ -47,26 +62,32 @@ int get_cpu_temperature(void)
     }
 
     // AMD CPU
-    else if (cpuid_info.vendor_id.str[0] == 'A' && cpuid_info.version.extendedFamily > 0 && cpuid_info.version.extendedFamily < 8) {
+    else if (cpuid_info.vendor_id.str[0] == 'A' && cpuid_info.version.family == 0xF) { // Target only K8 & newer
+
+        if(cpuid_info.version.extendedFamily >= 8) {        // Target Zen µarch and newer. Use SNM to get temperature.
```
Comment (debrouxl) [inline system/temperature.c:67]: Typos "if(" (also below), "SNM".

**Comment by debrouxl** [comment]
On my K8, the temperature reported by memtest86+ is "good", i.e. it lines with the stupid temperature range reported by `sensors` on all of my K8/K10/Bulldozer/Piledriver processors :)
P1110524
The first temperature displayed by memtest86+ was 12°C, i.e. below ambient temperature, probably 20-30s after the power up, and the computer wasn't fully cold (maybe a couple dozen minutes downtime). On the K10 (Athlon II X4 640, permanently down because of 1/2 damaged RAM slot), a temperature of ~44°C, according to `sensors`, meant overheating, with Machine Check Errors eventually creeping in every few months, until the computer was dusted.

**Comment by x86fr** [comment]
I saw the same behavior on some of my boards, by BIOS reports the same temperature. Thermal sensors on K8/K10 were a real mess and very unreliable 😒 

If you find something wrong, please open a separate issue for the specific CPU, with the exact CPUID (eg: there is 2 different Athlon 64 X2 5600+ 😢). 

I will merge this as it seems barely ok on all K8/K10 platforms I tried.

**Comment by jonathan-teh** [comment]
FWIW, this works fine on my K8 (dual-core 65nm Brisbane). Booting from cold I saw 21°C, a few degrees above ambient, briefly before it climbed rapidly.
memtest-a64x2
From Linux `/proc/cpuinfo`:
```
vendor_id       : AuthenticAMD
cpu family      : 15
model           : 107
model name      : AMD Athlon(tm) 64 X2 Dual Core Processor 5000+
stepping        : 2
```
and `sensors` after being left idle for a few minutes:
```
k8temp-pci-00c3
Adapter: PCI adapter
Core0 Temp:   +21.0°C  
Core0 Temp:   +20.0°C  
Core1 Temp:   +31.0°C  
Core1 Temp:   +22.0°C  
```

**Pull request merged**